gemrb/core/System/CString.h

```diff
@@ -0,0 +1,187 @@
+#include <cassert>
+#include <cctype>
+#include <cstdarg>
+#include <cstdint>
+#include <cstring>
+
+#ifndef WIN32
+# define stricmp strcasecmp
+# define strnicmp strncasecmp
+#endif
+
+namespace GemRB {
+
+#ifndef HAVE_STRNLEN
+GEM_EXPORT int strnlen(const char* string, int maxlen);
+#endif
+#ifndef HAVE_STRLCPY
+GEM_EXPORT size_t strlcpy(char *d, const char *s, size_t l);
+#endif
+
+#ifndef WIN32
+GEM_EXPORT char* strlwr(char* string);
+#endif
+
+GEM_EXPORT int strlench(const char* string, char ch);
+
+constexpr int NoTransform(int c) { return c; }
+
+template <typename STR_T, int(*TRANS)(int) = NoTransform>
+struct CstrHash
+{
+	size_t operator() (const STR_T &str) const {
+		size_t nHash = 0;
+		for (const auto& c : str) {
+			if (c == '\0')
+				break;
+			nHash = (nHash << 5) ^ TRANS(c);
+		}
+		return nHash;
+	}
+};
+
+template <typename STR_T>
+using CstrHashCI = CstrHash<STR_T, std::tolower>;
+
+template<size_t LEN, int(*CMP)(const char*, const char*, size_t) = strncmp>
+class FixedSizeString {
+	char str[LEN + 1] {'\0'};
+	
+public:
+	FixedSizeString() noexcept = default;
+	FixedSizeString(std::nullptr_t) noexcept = delete;
+	FixedSizeString& operator=(std::nullptr_t) noexcept = delete;
+	
+	FixedSizeString(const char* cstr) noexcept {
+		operator=(cstr);
+	}
+	
+	FixedSizeString& operator=(const char* c) noexcept {
+		if (c) {
+			strncpy(str, c, LEN);
+		} else {
+			std::fill(begin(), end(), '\0');
+		}
+		return *this;
+	}
+	
+	FixedSizeString(const FixedSizeString&) noexcept = default;
+	FixedSizeString& operator=(const FixedSizeString&) noexcept = default;
+	
+	template<typename T>
+	typename std::enable_if<std::is_integral<T>::value, char>::type
+	operator[](T i) const noexcept {
+		assert(i < static_cast<T>(LEN));
+		return str[i];
+	}
+	
+	template<typename T>
+	typename std::enable_if<std::is_integral<T>::value, char&>::type
+	operator[](T i) noexcept {
+		assert(i < static_cast<T>(LEN));
+		return str[i];
+	}
+	
+	bool operator==(const char* cstr) const noexcept {
+		return CMP(str, cstr, LEN) == 0;
+	}
+	
+	bool operator==(const FixedSizeString& other) const noexcept {
+		return CMP(str, other.str, LEN) == 0;
+	}
+	
+	bool operator!=(const char* cstr) const noexcept {
+		return CMP(str, cstr, LEN) != 0;
+	}
+	
+	bool operator!=(const FixedSizeString& other) const noexcept {
+		return CMP(str, other.str, LEN) != 0;
+	}
+	
+	bool operator<(const char* cstr) const noexcept {
+		return CMP(str, cstr, LEN) < 0;
+	}
+	
+	bool operator<(const FixedSizeString& other) const noexcept {
+		return CMP(str, other.str, LEN) < 0;
+	}
+	
+	bool StartsWith(const char* cstr, size_t len) const noexcept {
+		return CMP(str, cstr, len) == 0;
+	}
+	
+	uint8_t CStrLen() const noexcept {
+		return strnlen(str, sizeof(str));
+	}
+	
+	void Reset() noexcept {
+		std::fill(begin(), end(), '\0');
+	}
+	
+	void SNPrintF(const char* format, ...) noexcept {
+		va_list args;
+		va_start(args, format);
+		vsnprintf(str, sizeof(str), format, args);
```

Comment:
I see this error on MSYS2, maybe it needs `<cstdio>`:

```
D:/Sources/gemrb/gemrb/core/System/CString.h:148:26: error: 'vsnprintf' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
  148 |                 vsnprintf(str, sizeof(str), format, args);
      |                 ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from C:/MSYS2/mingw64/x86_64-w64-mingw32/include/locale.h:12,
                 from C:/MSYS2/mingw64/include/c++/11.2.0/clocale:42,
                 from C:/MSYS2/mingw64/include/c++/11.2.0/x86_64-w64-mingw32/bits/c++locale.h:41,
                 from C:/MSYS2/mingw64/include/c++/11.2.0/bits/localefwd.h:40,
                 from C:/MSYS2/mingw64/include/c++/11.2.0/string:43,
                 from D:/Sources/gemrb/gemrb/core/System/String.h:24,
                 from D:/Sources/gemrb/gemrb/includes/ie_types.h:33,
                 from D:/Sources/gemrb/gemrb/includes/globals.h:31,
                 from D:/Sources/gemrb/gemrb/core/Animation.h:26,
                 from D:/Sources/gemrb/gemrb/core/CharAnimations.h:27,
                 from D:/Sources/gemrb/gemrb/core/CharAnimations.cpp:21:
C:/MSYS2/mingw64/x86_64-w64-mingw32/include/stdio.h:464:5: note: 'int vsnprintf(char*, size_t, const char*, char*)' declared here, later in the translation unit
  464 | int vsnprintf (char *__stream, size_t __n, const char *__format, __builtin_va_list __local_argv)
      |     ^~~~~~~~~
```

Reply:
Confirming.

Reply:
@MarcelHB can you get a full list of changes required to build? when this is merged you can submit it as a PR. Ultimately, all this `printf` stuff should probably be rewritten with libfmt. I'll open some new tasks when this is in so we can bring everything together with the same formatting functionality.

Reply:
That's the only thing I encountered, but I can do this anyway if you like.